File: python/src/engine.py
class BacktestEngine:
    def __init__(self, initial_balance = 1000, fee = 0.001):
        self.initial_balance = initial_balance
        self.fee = fee
        self.position = None

    def run_backtest(self, df):
        data = df.copy()
        equity_history = []
        for index, row in data.iterrows():
            if self.position is None and row["buy_signal"]:
                open_price = row["close"]
                open_usd_size = self.initial_balance * (1 - self.fee)
                self.position = {
                    'open_price' : open_price,
                    'usd_size' : open_usd_size
                }
            elif self.position and row['sell_signal']:
                close_price = row["close"]
                trade_result = (close_price - self.position["open_price"]) / self.position['open_price']
                position_value_at_exit = self.position['usd_size'] * (1 + trade_result)
                self.initial_balance = position_value_at_exit * (1 - self.fee)
                self.position = None

            equity_history.append(self.initial_balance)
        
        data['equity'] = equity_history
        return data

File: cpp/src/Engine.cpp
#include "Engine.hpp"

Engine::Engine(double initial_balance, double fee): 
balance(initial_balance),initial_balance_ref(initial_balance), fee(fee){
    position.is_active = false;
}

std::vector<Candle>& Engine::run_backtester(std::vector<Candle> &dataset){
    for(int i = 0; i < dataset.size();i++){
        if(!position.is_active && dataset[i].buy_signal){
            position.open_price = dataset[i].close;
            position.usd_size = balance  * (1 - fee);
            position.is_active = true;
        }
        else if(position.is_active && dataset[i].sell_signal){
            double trade_result = (dataset[i].close - position.open_price) / position.open_price;
            double position_value_at_exit = position.usd_size * (1 + trade_result);
            balance = position_value_at_exit * (1 - fee);
            position.is_active = false;

        }

    }
    return dataset;
}

double Engine::getFinalBalance() const{
    return balance;
}
